### DISCUSSION TITLE
2.3.3: 分类下条目数量标注；#标签、嵌套标签支持正则表达式

### DISCUSSION THREAD

**Discussion by MuiseDestiny** [MuiseDestiny/zotero-style#191] [Q&A] [answered]
* 分类数量标注
* #标签排序，0-9A-Z
* #标签支持正则表达式，例如一个标签`#AAA/BBB/CCC`很长，只想显示CCC，可以用正则`/^#(?:.+\/)*(.+)/`，这个正则不仅用于#标签的原始标签->正则匹配的标签的映射，还作为嵌套标签的验证，属于高级用法，详见下面`getTagMatch`函数

**如果你不会正则表达式，但有高级需求，请在讨论中提问**

![image](https://user-images.githubusercontent.com/51939531/222413735-32a6fb5f-16f9-4643-8ea0-4a4c8801bd29.png)

```ts
static getTagMatch(tag: string) {
    // 监测是否为正则表达式
    const rawString = Zotero.Prefs.get(`${config.addonRef}.textTagsColumn.match`) as string
    const res = rawString.match(/\/(.+)\/(\w*)/)
    let regex: RegExp;
    // 是正则表达式
    if (res) {
      regex = new RegExp(res[1], res[2])
    }
    // 不以xxx开头
    else if (rawString.startsWith("~~")) {
      regex = new RegExp(`^([^${rawString.slice(2)}].+)`)
    }
    // 以xxx开头
    else {
      regex = new RegExp(`^${rawString}(.+)`)
    }
    const matched = tag.match(regex)
    return (matched && matched.slice(1).join("")) || ""
  }
```

<hr /><em>This discussion was created from the release <a href='https://github.com/MuiseDestiny/zotero-style/releases/tag/2.3.3'>2.3.3: 分类下条目数量标注；#标签、嵌套标签支持正则表达式</a>.</em>

**Comment by felicity1559** [comment]


**Reply by MuiseDestiny** [reply]
好像安装的不是最新，在github下载安装试试。

**Reply by MuiseDestiny** [reply]
/^#(?:.+\/)*(.+)/

然后输入上面这一串就可以了。

**Comment by ghost** [comment, marked as answer]
请问如何关闭分类下条目数量标注的功能，看得我有点难受

**Reply by MuiseDestiny** [reply]
使用旧版

**Reply by ghost** [reply]
后期也不会有开和关的选项吗

**Reply by MuiseDestiny** [reply]
或者等提供关闭入口

**Reply by MuiseDestiny** [reply]
会啊，得慢慢写

**Reply by ghost** [reply]
好的，谢谢！还有一个建议就是，我觉得随着功能越来越多，可以在首选项或者工具中写个菜单设置，更方便管理。

**Reply by MuiseDestiny** [reply]
是的，但写那个比较费时间，而且关闭开启功能不会很频繁先这么关把

**Reply by MuiseDestiny** [reply]
![image](https://user-images.githubusercontent.com/51939531/222869840-ab78e573-3b9a-47de-9acc-c2c27593dc45.png)

搜索，addNumberToCollectionTree，false，重启即可

**Reply by ghost** [reply]
感谢

**Reply by ghost** [reply]
不懂就问，请问这个addNumberToCollectionTree哪儿搜

**Reply by MuiseDestiny** [reply]
![image](https://user-images.githubusercontent.com/51939531/222870526-d80c2454-1806-43a4-b488-11bb6a105a1d.png)

**Reply by MuiseDestiny** [reply]
嵌套标签也有数字，function.Tags也可以关

**Reply by MuiseDestiny** [reply]
![image](https://user-images.githubusercontent.com/51939531/222870976-f73aeafc-1374-4d0b-81e5-7a837be5610d.png)

**Comment by ghost** [comment]
<img width="500" alt="WeChat30f2b80e09ce69df58290f18759c7a6c" src="https://user-images.githubusercontent.com/44020658/222872223-29fe0b52-3e4b-497b-8d00-e3a640cc626b.png">
<img width="342" alt="WeChatfb2e5ba2b9134f5cae2863db7417b0a4" src="https://user-images.githubusercontent.com/44020658/222872229-f654511e-5988-41a6-a6d2-d38ea5eff03f.png">

为什么这么设置了，SCIIF（5）这个地方还是不显示IF（5）？

**Comment by qiuqiu0122** [comment]


**Reply by MuiseDestiny** [reply]
#标签 的列设置里的Match怎么设置的。可能跟那个有关系

**Reply by qiuqiu0122** [reply]
![image](https://user-images.githubusercontent.com/127204848/223423014-c6ae671b-e094-49e4-a37d-c30d9837a463.png)
您好，我之前没有设置这个，现在设置了一个#，这样可以吗？

**Reply by qiuqiu0122** [reply]
![image](https://user-images.githubusercontent.com/127204848/223423241-f7add826-5292-42b6-b382-d94038f2b440.png)
但是我发现左下角这里没有了三个图标，也不能分级

**Reply by MuiseDestiny** [reply]
重启呢

**Reply by qiuqiu0122** [reply]
> 重启呢

重启也没有诶

**Reply by qiuqiu0122** [reply]
> 重启呢

之前是显示不出部分标签，然后忽然这个功能都显示不出来了

**Reply by qiuqiu0122** [reply]
![image](https://user-images.githubusercontent.com/127204848/223435091-f4677508-fa2c-465e-9a8c-039f15eedc62.png)
这个是首选项的情况，我选择了这些，不知道为什么会没有嵌套标签，很喜欢这个理念

**Reply by MuiseDestiny** [reply]
我试试哦

**Reply by MuiseDestiny** [reply]
我开启和你一样的功能，嵌套标签是正常的

**Reply by qiuqiu0122** [reply]
> 我开启和你一样的功能，嵌套标签是正常的

呜呜，那怎么办呀，遗憾不能用这个功能了。重启了好几次电脑了。

**Reply by MuiseDestiny** [reply]
> > 我开启和你一样的功能，嵌套标签是正常的
> 
> 呜呜，那怎么办呀，遗憾不能用这个功能了。重启了好几次电脑了。

要不然功能全开启试试，或者禁用其它插件（但我觉得应该没有冲突）。

然后还不行，贴图下输出日志（帮助-调试日志-查看输出日志）

**Reply by qiuqiu0122** [reply]
> > > 我开启和你一样的功能，嵌套标签是正常的
> > 
> > 
> > 呜呜，那怎么办呀，遗憾不能用这个功能了。重启了好几次电脑了。
> 
> 要不然功能全开启试试，或者禁用其它插件（但我觉得应该没有冲突）。
> 
> 然后还不行，贴图下输出日志（帮助-调试日志-查看输出日志）

您好,我试了一下还是不行，我把输出日志贴出来了，麻烦您了，十分感谢！！

**Reply by MuiseDestiny** [reply]
先禁用其他插件重启zotero也试试

**Reply by qiuqiu0122** [reply]
> 先禁用其他插件重启zotero也试试

除了zotero style我把所有都禁了，还是不行

**Reply by qiuqiu0122** [reply]
![image](https://user-images.githubusercontent.com/127204848/223448460-547d2dfa-2678-451f-90c1-1a2d06be95a5.png)

**Comment by qiuqiu0122** [comment]


**Reply by MuiseDestiny** [reply]
这里没有任何报错

**Comment by ayden19** [comment]
嵌套标签是例如：#材料/无机/Ni基。想在#标签一栏中，分别显示材料、无机、Ni基三个，而不是显示成材料/无机/Ni基，应该怎么做？

**Comment by shandianchengzi** [comment]
这个功能真的超级赞！但是嵌套标签不能支持群组文库，无法推广到团队里使用QAQ，希望能添加群组支持

**Comment by shaoqx** [comment]
发现在最新版本（7.0.27 (64-bit)，插件版本5.7.6.6）中如果正则表达式写成

`/^[^/#](?:.+/)+(.+)/`

左下角的嵌套标签会显示空白，同时#标签正常

我这么写是由于#开头的标签如果被同时设置为颜色标签，无法在“标签”列显示（也就是不显示彩色圆点），于是尝试定义字符串内部所有包含`/`的标签作为嵌套标签，但是发现上面的问题。

如果设置为

`/^[^/#](.+/)+(.+)/`

左下角显示正常，但是#标签一列显示会缺少第一个字符

**Reply by shaoqx** [reply]
发现用@代替#貌似能解决我的所有问题，颜色标签显示正常了

**Reply by shaoqx** [reply]
不对 左下角还是不显示
